### ISSUE TITLE
Change in exported data in Carto3 v7

### ISSUE THREAD

**Issue by stephan1312** [stephan1312/SlicerEAMapReader#4]
Since I will probably not find the time to address this in the near future, I would like to document some open issue here.

The data exported from the CARTO 3 workstation has changed somewhat from v6 to v7:
In v6 and earlier, the .mesh file for each map contained the anatomy (triangulated mesh) and the interpolated electric data for each mesh vetex.
In v7, the .mesh file only contains the triangulated mesh, but there is no more vertex-wise interpolated electric information exported

Background: During mapping, the map surface is created as the (smoothed) outer shell of all points in space where the catheter has been. At the same time, a point cloud is created where a point in space is recorded at the catheter electrode location with each acquired beat. These points and their electric data are projected onto the surface (nominal setting: use only points within 7 mm from the surface). The electric data is then interpolated between the surface-projected points. This  point cloud is exported as ????_car.txt and is already read by the plugin into Slicer as a markup fiducial points set. 

Issue: The electric data interpolation performed by the workstation is no longer exported. We therefore need to re-calculate the interpolation at import. The necessary data should be available (point cloud with electric data and surface mesh without). However, interpolation settings (maximum point distance from surface, maximum interpolation fill distance) on the workstation change the map appearance although the underlying data is the same. These setting should probably be changeable by the user in EAMapReader, too.

**Comment by Jsalas424** [comment]
I have emailed Dr. Veltmann about potentially collaborating to add V7 compatibility.

**Comment by stephan1312** [comment]
I have talked to Dr. Veltmann yesterday and told him about your going to contact him, but he has not received an e-mail.

Anyway, you can always e-mail me at hohmann.stephan@mh-hannover.de

 

I'd love to collaborate with you to add v7 compatibility.

 

Best regards

 

Stephan

 
 

Gesendet: Montag, 30. November 2020 um 22:20 Uhr
Von: "Jsalas424" <notifications@github.com>
An: "stephan1312/SlicerEAMapReader" <SlicerEAMapReader@noreply.github.com>
Betreff: Re: [stephan1312/SlicerEAMapReader] Change in exported data in Carto3 v7 (#4)


 

I have emailed Dr. Veltmann about potentially collaborating to add V7 compatibility.

**Comment by Mazze** [comment]
Has there been any attempt to implement the interpolation? I will need the interpolation, while I will not need Slicer I plan on using python. Thus, I could see me contributing on the interpolation side, but not the GUI side.

**Comment by stephan1312** [comment]
I am not aware of any attempt to implement the Carto3 v7 interpolation. I have played with some maps on the workstation to deduce & confirm some properties of the interpolation algorithm, but that's mostly it so far. If you would like to contribute, please feel free.

**Comment by Mazze** [comment]
I created a simple standalone script for the interpolation of the LAT to the mesh. So far, I have not had the time to adapt it for your plugin. I hope to get that done soon. The script can be found [here](https://github.com/Mazze/CartoInterpolation/) 

I think the exported points, are already filtered by the distance to the surface. Meaning that if the standard of 7 units is changed in carto that should be reflected in the data. I came to the conclusion because non of my data point was more than 7units from the surface, but 7 was present.

**Comment by javierfuster** [comment]
Hello, 

Anyone knows where I can find in the CARTO3 v7 files the information about the LAT? Thank you

**Comment by Nisal47** [comment]
Hi

Could you please tell me how to map interpolation data into colors?

**Comment by Mazze** [comment]
> Hi
> 
> Could you please tell me how to map interpolation data into colors?

Can you be more specific in what you like to do? The interpolated values will be some single value, thus the normal solution is to create a lookup table. I am pretty sure this is not the answer you are looking for, but your question is missing context.

**Comment by Nisal47** [comment]
Yes. That is the kind of answer I am looking for. I am implementing Carto file reader for one of my projects. Also, I am going to use a kind of lookup table to map those interpolated values. As there are three types of values, which one I should use (Unipolar, Bipolar, LAT) for coloring? And, if you have an interpolation-color lookup table, can share it with me?

**Comment by Nisal47** [comment]
> > Hi
> > Could you please tell me how to map interpolation data into colors?
> 
> Can you be more specific in what you like to do? The interpolated values will be some single value, thus the normal solution is to create a lookup table. I am pretty sure this is not the answer you are looking for, but your question is missing context.

Yes. That is the kind of answer I am looking for. I am implementing Carto file reader for one of my projects. Also, I am going to use a kind of lookup table to map those interpolated values. As there are three types of values, which one I should use (Unipolar, Bipolar, LAT) for coloring? And, if you have an interpolation-color lookup table, can share it with me?

**Comment by Mazze** [comment]
I can not tell you which one (Unipolar, Bipolar, LAT) you need to visualize. The Unipolar and Bipolar are voltage measurements from the catheter and can tell you about the tissue mapped. For example, you could use low voltage areas to identify disease tissue. The LAT is the local activation time in ms, thus giving the spread of activation. Which is the correct one to show depending on your problem. 
Most visualization packages come with their own lookup table. Therefore, I have no table on hand, but if you need to create your own you can define a pathway in the RBG space and sample it as often as you like.

**Comment by Nisal47** [comment]
> I can not tell you which one (Unipolar, Bipolar, LAT) you need to visualize. The Unipolar and Bipolar are voltage measurements from the catheter and can tell you about the tissue mapped. For example, you could use low voltage areas to identify disease tissue. The LAT is the local activation time in ms, thus giving the spread of activation. Which is the correct one to show depending on your problem. Most visualization packages come with their own lookup table. Therefore, I have no table on hand, but if you need to create your own you can define a pathway in the RBG space and sample it as often as you like.

Thank you!!

**Comment by stephan1312** [comment]
> I created a simple standalone script for the interpolation of the LAT to the mesh. So far, I have not had the time to adapt it for your plugin. I hope to get that done soon. The script can be found [here](https://github.com/Mazze/CartoInterpolation/)
> 
> I think the exported points, are already filtered by the distance to the surface. Meaning that if the standard of 7 units is changed in carto that should be reflected in the data. I came to the conclusion because non of my data point was more than 7units from the surface, but 7 was present.



> I created a simple standalone script for the interpolation of the LAT to the mesh. So far, I have not had the time to adapt it for your plugin. I hope to get that done soon. The script can be found [here](https://github.com/Mazze/CartoInterpolation/)
> 
> I think the exported points, are already filtered by the distance to the surface. Meaning that if the standard of 7 units is changed in carto that should be reflected in the data. I came to the conclusion because non of my data point was more than 7units from the surface, but 7 was present.


Hi Mazze,
I somehow missed your post back then and just saw it today. The script looks great. I will certainly test it with a couple of maps soon.
Thank you!

**Comment by borantolic** [comment]
> I can not tell you which one (Unipolar, Bipolar, LAT) you need to visualize. The Unipolar and Bipolar are voltage measurements from the catheter and can tell you about the tissue mapped. For example, you could use low voltage areas to identify disease tissue. The LAT is the local activation time in ms, thus giving the spread of activation. Which is the correct one to show depending on your problem. Most visualization packages come with their own lookup table. Therefore, I have no table on hand, but if you need to create your own you can define a pathway in the RBG space and sample it as often as you like.

Hi! Is there any chance you could give me some guidance on how to use your script? I am a cardia electrophysiologist, but not very good at figuring how scripts work by myself. I would like to create bipolar and unipolar maps from exported CARTO files. Furthermore it would be best if I could define color thresholds - I guess this is done with these color lookup tables - but can you give me guidance on that to? Or direct me to a webpage where this is explained? 
Thanks!
Bor Antolic

**Comment by stephan1312** [comment]
> > I can not tell you which one (Unipolar, Bipolar, LAT) you need to visualize. The Unipolar and Bipolar are voltage measurements from the catheter and can tell you about the tissue mapped. For example, you could use low voltage areas to identify disease tissue. The LAT is the local activation time in ms, thus giving the spread of activation. Which is the correct one to show depending on your problem. Most visualization packages come with their own lookup table. Therefore, I have no table on hand, but if you need to create your own you can define a pathway in the RBG space and sample it as often as you like.
> 
> Hi! Is there any chance you could give me some guidance on how to use your script? I am a cardia electrophysiologist, but not very good at figuring how scripts work by myself. I would like to create bipolar and unipolar maps from exported CARTO files. Furthermore it would be best if I could define color thresholds - I guess this is done with these color lookup tables - but can you give me guidance on that to? Or direct me to a webpage where this is explained? Thanks! Bor Antolic

You might want to check out this thread on the 3D Slicer discourse forum: 
https://discourse.slicer.org/t/how-to-register-cardiac-ea-map-surface-mesh-with-3d-dicom-image/26121/7
